src/adapter/src/coord/sequencer/inner.rs

```diff
@@ -2116,67 +2116,99 @@ impl Coordinator {
-            self.check_txn_timedomain_conflicts(
-                session,
-                &determination.timestamp_context,
-                id_bundle,
-            )?;
-
-            // If we've already acquired read holds for the txn, we can skip doing so again
-            if !self.txn_reads.contains_key(session.conn_id()) {
-                if let Some(timestamp) = determination.timestamp_context.timestamp() {
-                    let timedomain_id_bundle = self.timedomain_for(
-                        source_ids,
-                        &timeline_context,
-                        session.conn_id(),
-                        cluster_id,
-                    )?;
-
-                    let read_holds =
-                        self.acquire_read_holds(timestamp.clone(), &timedomain_id_bundle);
+            // Either set the valid read ids for this transaction (if it's the first statement in a
+            // transaction) otherwise verify the ids referenced in this query are in the timedomain.
+            if let Some(txn_reads) = self.txn_reads.get(session.conn_id()) {
+                // Find referenced ids not in the read hold. A reference could be caused by a
+                // user specifying an object in a different schema than the first query. An
+                // index could be caused by a CREATE INDEX after the transaction started.
+                let allowed_id_bundle = txn_reads.id_bundle();
+                let outside = source_bundle.difference(&allowed_id_bundle);
+                // Queries without a timestamp and timeline can belong to any existing timedomain.
+                if determination.timestamp_context.contains_timestamp() && !outside.is_empty() {
+                    let valid_names =
+                        self.resolve_collection_id_bundle_names(session, &allowed_id_bundle);
+                    let invalid_names = self.resolve_collection_id_bundle_names(session, &outside);
+                    return Err(AdapterError::RelationOutsideTimeDomain {
+                        relations: invalid_names,
+                        names: valid_names,
+                    });
+                }
+            } else {
+                if let Some((timestamp, bundle)) = potential_read_holds {
+                    let read_holds = self.acquire_read_holds(timestamp, bundle);
                     self.txn_reads.insert(session.conn_id().clone(), read_holds);
                 }
             }
         }
 
+        let mut transaction_determination = determination.clone();
+        if !when.is_transactional() {
+            // If the query uses AS OF, ignore the timestamp so that this query can be executed in
+            // this transaction and also to prevent future statements from being locked to this
+            // timestamp.
+            transaction_determination.timestamp_context = TimestampContext::NoTimestamp;
+        };
+        session.add_transaction_ops(TransactionOps::Peeks(transaction_determination))?;
```

Comment:
Previously we would not call this function if 

`!when.is_transactional() && !matches!(session.transaction(), &TransactionStatus::InTransaction(_))`.

Now we do call this function in this scenario, which is, I believe, why the test are now failing in `transaction.slt`. You're change seems like the correct behavior, but we did seem to intentionally not call this function in implicit transactions when the query was `AS OF` and I can't remember why.

Reply:
I pushed two commits which fix the tests. The test failure happened because implicit transactions used to skip this call when the current query used `AS OF`. which means that implicit transactions allowed the following scenarios:

- Timestamp independent peeks in a subscribe transaction.
- Timestamp independent subscribes in a peek transaction.

However, these weren't allowed in explicit transactions, because we'd call this function in explicit transactions.


With your changes, we always call this function, so those scenarios were not allowed in both implicit transactions and explicit transactions. The tests were testing theses scenarios in implicit transactions so they started to fail.

My commits update `add_transaction_ops` to allow the mentioned scenarios in both implicit transactions and explicit transactions. This seems more correct that only calling `add_transaction_ops` for explicit transaction. Though the old behavior seemed so intentional that it's a bit confusing.

Reply:
As per the our discussion over slack, it's probably best to revert my commits and update this so `add_transaction_ops` is not called for implicit transactions, which matches the old behavior. We can then address the differences between implicit and explicit transactions in a followup PR.

Reply:
Reverted this code to what it was before and added a todo.